File: 18.kt
fun main() {
     val up = Pair(0, -1); val right = Pair(1, 0); val down = Pair(0, 1); val left = Pair(-1, 0)

    fun move(pos: Pair<Int, Int>, dir: Pair<Int, Int>, n: Int): Pair<Int, Int> = Pair(pos.first + dir.first * n, pos.second + dir.second * n)

    fun result(len: Long, corners: List<Pair<Int, Int>>): Long {
        var result: Long = 0
        for (i in 1..<(corners.size - 1))
            result += corners[i].first.toLong() * (corners[i + 1].second.toLong() - corners[i - 1].second.toLong())
        return result / 2 + len / 2 + 1
    }

    var pos = Pair(0, 0); var pos2 = Pair(0, 0)
    val corners = ArrayList<Pair<Int, Int>>(); val corners2 = ArrayList<Pair<Int, Int>>()
    var len1 = 0; var len2: Long = 0
    corners.add(Pair(0, 0))
    corners2.add(Pair(0, 0))
    do {
        val l = readlnOrNull() ?: break;
        val line = l.split(' ')
        val n = line[1].toInt()
        when (line[0][0]) {
            'U' -> pos = move(pos, up, n); 'D' -> pos = move(pos, down, n); 'L' -> pos = move(pos, left, n); 'R' -> pos = move(pos, right, n)
        }
        corners.add(pos)
        len1 += n
        val n2 = line[2].substring(2, 7).toInt(16)
        when (line[2][7]) {
            '3' -> pos2 = move(pos2, up, n2); '1' -> pos2 = move(pos2, down, n2); '2' -> pos2 = move(pos2, left, n2); '0' -> pos2 = move(pos2, right, n2)
        }
        len2 += n2
        corners2.add(pos2)
    } while (true)
    println(listOf(result(len1.toLong(), corners), result(len2, corners2)))
}

File: 18.go
package main

import "bufio"
import "fmt"
import "os"
import "strconv"
import "strings"

var directions = map[string]struct{dx, dy int64}{
	"D": {0, 1}, "U": {0, -1}, "R": {1, 0}, "L": {-1, 0},
}
var part2dirs = map[uint8]struct{dx, dy int64}{
	'3': {0, -1}, '1': {0, 1}, '2': {-1, 0}, '0': {1, 0},
}

type pos struct {x, y int64}

func area(corners []pos, l int64) int64 {
	var result int64
	for i := 1; i < len(corners) - 1; i++ {
		result += corners[i].x * (corners[i + 1].y - corners[i - 1].y)
	}
	return result / 2 + l / 2 + 1
}

func main() {
	var scanner = bufio.NewScanner(os.Stdin)
	var part1corners = []pos{{0, 0}}
	var part2corners = []pos{{0, 0}}
	var pos1, pos2, len1, len2 = pos{0, 0}, pos{0, 0}, int64(0), int64(0)
	for scanner.Scan() {
		var items = strings.Split(scanner.Text(), " ")
		{
		var dir, found = directions[items[0]]
		if !found { panic("invalid directions " + items[0]) }
		n, _ := strconv.ParseInt(items[1], 10, 32)
		pos1 = pos{pos1.x + dir.dx * n, pos1.y + dir.dy * n}
		part1corners = append(part1corners, pos1)
		len1 += n
		}
		{
		var dir, found = part2dirs[items[2][7]]
		if !found { panic("invalid directions " + items[0]) }
		n, _ := strconv.ParseInt(items[2][2:7], 16, 32)
		len2 += n
		pos2 = pos{pos2.x + dir.dx * n, pos2.y + dir.dy * n}
		part2corners = append(part2corners, pos2)
		}
	}
	fmt.Println(area(part1corners, len1), area(part2corners, len2))
}
